### PR TITLE
build: Speed up build by separating construct languages

### PR DESCRIPTION
**Pull request by gliptak**
Signed-off-by: Gábor Lipták <gliptak@gmail.com>

Fixes # 

Time decreased from 20+ minutes to 10 minutes

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -3,11 +3,38 @@ name: Build Pull Requests
 on: [pull_request]
```
Comment (ryparker) [inline .github/workflows/build.yml:3]: This is a great optimization. I think we can do even better too. If we create a workflow file for each job we can take advantage of [triggering the specific build depending on which files were changed](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths).

By the way can we rename the `npm` job and bash script to `typescript`? This way the names match the dir they're under.


Example:

```yaml
# /.github/workflows/build-typescript.yml

on:
  pull_request:
    paths:
      - typescript/**

jobs:
  build-typescript:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Runs docker build with JSII superchain
      run: docker run --rm --net=host -t -v $PWD:$PWD -w $PWD jsii/superchain /bin/bash -c "scripts/build-typescript.sh"
```

Mergify will need to be updated as well. By using [Mergify's conditions syntax](https://docs.mergify.io/conditions) the following should work:

```yaml
# ./.mergify.yml
# See https://doc.mergify.io

pull_request_rules:
  - name: automatic merge
    actions:
      comment:
        message: Thank you for contributing! Your pull request is now being automatically merged.
      merge:
        strict: smart
        method: squash
        strict_method: merge
      delete_head_branch: {}
      dismiss_reviews: {}
    conditions:
      - -title~=(WIP|wip)
      - -label~=^(blocked|do-not-merge)
      - -merged
      - -closed
      - "#approved-reviews-by>=1"
      - -approved-reviews-by~=author
      - "#changes-requested-reviews-by=0"
      - status-success=Semantic Pull Request
      - and:
        - or: # if typescript files have changed, wait for build
          - files~=^typescript/
          - status-success=build-typescript
        - or: # if typescript files did not change, continue
          - -files~=^typescript/

      - and:
        - or: # if java files have changed, wait for build
          - files~=^java/
          - status-success=build-java
        - or: # if java files did not change, continue
          - -files~=^java/

      - and:
        - or: # if python files have changed, wait for build
          - files~=^python/
          - status-success=build-python
        - or: # if python files did not change, continue
          - -files~=^python/

      - and:
        - or: # if csharp files have changed, wait for build
          - files~=^csharp/
          - status-success=build-csharp
        - or: # if csharp files did not change, continue
          - -files~=^csharp/

  - name: remove stale reviews
    actions:
      dismiss_reviews: {}
    conditions:
      - base=master
  - name: if fails conventional commits
    actions:
      comment:
        message: Title does not follow the guidelines of [Conventional Commits](https://www.conventionalcommits.org). Please adjust title before merge.
    conditions:
      - -status-success=Semantic Pull Request
```

Comment (gliptak) [reply .github/workflows/build.yml]: implemented

**Comment by gliptak** [comment]
@ryparker not quite sure how this is to be tested

**Review by ryparker — DISMISSED** [review]
Looks great to me. I think the only way to test this is to merge to the main branch. If it doesn't work we can always revert and assess.

**Review by ryparker — APPROVED** [review]

**Comment by ryparker** [comment]
@Mergifyio refresh

**Pull request merged**